File: api/sync/src/main/java/com/coditory/sherlock/test/SherlockStub.java
package com.coditory.sherlock.test;

import com.coditory.sherlock.DistributedLock;
import com.coditory.sherlock.DistributedLockBuilder;
import com.coditory.sherlock.Sherlock;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

import static com.coditory.sherlock.Preconditions.expectNonNull;

/**
 * Used to stub {@link Sherlock} in tests.
 */
public final class SherlockStub implements Sherlock {
    private final Map<String, DistributedLock> locksById = new HashMap<>();
    private boolean defaultLockResult = true;

    /**
     * Make the stub produce released locks by default
     *
     * @return the stub instance
     */
    @NotNull
    public static SherlockStub withReleasedLocks() {
        return new SherlockStub()
            .withDefaultAcquireResult(true);
    }

    /**
     * Make the stub produce acquired locks by default
     *
     * @return the stub instance
     */
    @NotNull
    public static SherlockStub withAcquiredLocks() {
        return new SherlockStub()
            .withDefaultAcquireResult(false);
    }

    /**
     * Make the stub produce return a predefined lock.
     *
     * @param lock returned when creating a lock with the same id
     * @return the stub instance
     */
    @NotNull
    public SherlockStub withLock(@NotNull DistributedLock lock) {
        expectNonNull(lock, "lock");
        this.locksById.put(lock.getId(), lock);
        return this;
    }

    private SherlockStub withDefaultAcquireResult(boolean result) {
        this.defaultLockResult = result;
        return this;
    }

    @Override
    public void initialize() {
        // deliberately empty, nothing to initialize
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createLock() {
        return getLockOrDefault();
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createReentrantLock() {
        return getLockOrDefault();
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createOverridingLock() {
        return getLockOrDefault();
    }

    @Override
    public boolean forceReleaseAllLocks() {
        // deliberately empty
        return false;
    }

    private DistributedLockBuilder<DistributedLock> getLockOrDefault() {
        return new DistributedLockBuilder<>(this::getLockOrDefault);
    }

    private DistributedLock getLockOrDefault(String id, Duration duration, String ownerId) {
        DistributedLockMock defaultLock = DistributedLockMock.lockStub(id, defaultLockResult);
        return locksById.getOrDefault(id, defaultLock);
    }
}

File: api/coroutines/src/main/java/com/coditory/sherlock/coroutines/test/SherlockStub.kt
package com.coditory.sherlock.coroutines.test

import com.coditory.sherlock.DistributedLockBuilder
import com.coditory.sherlock.coroutines.DistributedLock
import com.coditory.sherlock.coroutines.Sherlock
import java.time.Duration

/**
 * Use to stub [Sherlock] in tests.
 */
class SherlockStub : Sherlock {
    private val locksById: MutableMap<String, DistributedLock> = HashMap()
    private var defaultLockResult = true

    /**
     * Make the stub produce return a predefined lock.
     *
     * @param lock returned when creating a lock with the same id
     * @return the instance
     */
    fun withLock(lock: DistributedLock): SherlockStub {
        locksById[lock.id] = lock
        return this
    }

    private fun withDefaultAcquireResult(result: Boolean): SherlockStub {
        defaultLockResult = result
        return this
    }

    override suspend fun initialize() {
        // deliberately empty
    }

    override fun createLock(): DistributedLockBuilder<DistributedLock> {
        return getLockOrDefault()
    }

    override fun createReentrantLock(): DistributedLockBuilder<DistributedLock> {
        return getLockOrDefault()
    }

    override fun createOverridingLock(): DistributedLockBuilder<DistributedLock> {
        return getLockOrDefault()
    }

    override suspend fun forceReleaseAllLocks(): Boolean {
        return false
    }

    private fun getLockOrDefault(): DistributedLockBuilder<DistributedLock> {
        return DistributedLockBuilder { id: String, _: Duration?, _: String ->
            val defaultLock = DistributedLockMock.lockStub(id, defaultLockResult)
            locksById.getOrDefault(id, defaultLock)
        }
    }

    companion object {
        /**
         * Make the stub produce released locks by default
         *
         * @return the instance
         */
        fun withReleasedLocks(): SherlockStub {
            return SherlockStub()
                .withDefaultAcquireResult(true)
        }

        /**
         * Make the stub produce acquired locks by default
         *
         * @return the instance
         */
        fun withAcquiredLocks(): SherlockStub {
            return SherlockStub()
                .withDefaultAcquireResult(false)
        }
    }
}
